modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Traverse.kt

```diff
@@ -5,11 +5,519 @@ import arrow.core.Id
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ * import arrow.core.extensions.either.applicative.applicative
+ * import arrow.core.k
+ * import arrow.core.ValidatedNel
+ * import arrow.core.extensions.either.foldable.isEmpty
+ * import arrow.core.invalidNel
+ * import arrow.core.validNel
+ *
+ * fun parseIntEither(s: String): Either<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
+ *   else Either.Left(NumberFormatException("$s is not a valid integer."))
+ *
+ * fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
+ *   else NumberFormatException("$s is not a valid integer.").invalidNel()
+ *
+ * fun main() {
+ *   //sampleStart
+ *   val list = listOf("1", "2", "3").k().traverse(Either.applicative(), ::parseIntEither)
+ *   val failFastList = listOf("1", "abc", "3", "4s").k().traverse(Either.applicative(), ::parseIntEither)
+ *   //sampleEnd
+ *   println("list= $list")
+ *   println("failFastList= $failFastList")
+ * }
+ * ```
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ * import arrow.core.Nel
+ * import arrow.core.ValidatedNel
+ * import arrow.core.extensions.nonemptylist.semigroup.semigroup
+ * import arrow.core.extensions.validated.applicative.applicative
+ * import arrow.core.fix
+ * import arrow.core.invalidNel
+ * import arrow.core.k
+ * import arrow.core.validNel
+ *
+ * fun parseIntEither(s: String): Either<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
+ *   else Either.Left(NumberFormatException("$s is not a valid integer."))
+ *
+ * fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
+ *   else NumberFormatException("$s is not a valid integer.").invalidNel()
+ *
+ * fun main() {
+ *   //sampleStart
+ *   val validatedList = listOf("1", "22w", "3", "33s").k()
+ *     .traverse(ValidatedNel.applicative(Nel.semigroup<NumberFormatException>()), ::parseIntValidated).fix()
+ *   //sampleEnd
+ *   println(validatedList)
+ * }
+ * ```
+ *
+ * Notice that in the [Either] case, should any string fail to parse the entire [traverse] is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option`). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of [traverse] is closely tied with the [Applicative] behavior of the data type, where computations are run in isolation.
+ *
+ * Going back to our `IO` example from the beginning with concurrency in mind, we can get an [Applicative] instance for `IO`, that will not short-circuit, by using `parApplicative`.
```

Comment:
```suggestion
 * Going back to our `IO` example from the beginning with concurrency in mind, we can get an [Applicative] instance for `IO`, that traverses concurrently, by using `parApplicative`. This is in contrast to usual `IO.applicative()` which executes in sequence.
```
Don't think short-circuit is whats important at this point. I would however stretch the differences between this and the default applicative for IO

Reply:
This is explained in a few lines down below

Reply:
I know, the point was not that it is not explained or wrong, but that it is not the important part. It is also not correct, it will short-circuit, but it can run effects from later elements before.
`listOf(1,2,3,4).parTraverse(IO { println(it); throw Error() })` can print anything from 1 to 4 or even all of it depending on who wins the race, but the result will be the first error encountered. So it short circuits as soon as it encounters an error, but because the effects start concurrently it can execute the effects of other items before doing so. The result is the same, except for side-effects

Reply:
Makes sense. I think the part with `short-circuit`, especially in regards to side-effects could be a great addition to the parTraverse Docs.
I hope I could nail your point.